app/src/main/java/io/appium/settings/RecorderThread.java

```diff
@@ -0,0 +1,437 @@
+
+            MediaFormat videoEncoderFormat = initVideoEncoderFormat(videoMime,
+                    finalVideoWidth, finalVideoHeight, videoBitrate);
+
+            videoEncoder.configure(videoEncoderFormat, null, null,
+                    MediaCodec.CONFIGURE_FLAG_ENCODE);
+            surface = videoEncoder.createInputSurface();
+            videoEncoder.start();
+
+            Handler handler = new Handler();
+            virtualDisplay = initVirtualDisplay(this.mediaProjection, surface, handler,
+                    finalVideoWidth, finalVideoHeight);
+
+            int sampleRate = RecorderConstant.AUDIO_CODEC_SAMPLE_RATE_HZ;
+            audioEncoder = initAudioCodec(sampleRate);
+            audioEncoder.start();
+
+            audioRecord = initAudioRecord(this.mediaProjection, sampleRate);
+
+            muxer = new MediaMuxer(this.outputFilePath,
+                    MediaMuxer.OutputFormat.MUXER_OUTPUT_MPEG_4);
+
+            // set output file orientation info
+            // note: this method must be run before muxer.start()
+            muxer.setOrientationHint(recordingRotation);
+
+            audioRecordThread = startAudioRecord(audioEncoder, audioRecord);
+            audioRecordThread.start();
+
+            MediaCodec.BufferInfo bufferInfo = new MediaCodec.BufferInfo();
+            lastAudioTimestampUs = NO_TIMESTAMP_SET;
+
+            while (!stopped && !hasAsyncError) {
+                if (!writeAudioBufferToFile(audioEncoder, muxer, bufferInfo)) {
+                    break;
+                }
+
+                if (videoTrackIndex >= 0 && audioTrackIndex < 0) {
+                    continue; // wait for audio config before processing any video data frames
+                }
+
+                if (!writeVideoBufferToFile(videoEncoder, muxer, bufferInfo)) {
+                    break;
+                }
+            }
+        } catch (Exception mainException) {
+            mainException.printStackTrace();
+        } finally {
+            if (muxer != null) {
+                muxer.stop();
+                muxer.release();
+                muxer = null;
+            }
+
+            if (virtualDisplay != null) {
+                virtualDisplay.release();
+                virtualDisplay = null;
+            }
+
+            if (surface != null) {
+                surface.release();
+                surface = null;
+            }
+
+            if (videoEncoder != null) {
+                videoEncoder.stop();
+                videoEncoder.release();
+                videoEncoder = null;
+            }
+
+            if (audioRecordThread != null) {
+                audioStopped = true;
+                try {
+                    audioRecordThread.join();
+                    audioRecordThread = null;
+                } catch (InterruptedException e) {
+                    Log.w(TAG, "Error releasing resources, audioRecordThread: ", e);
+                    e.printStackTrace();
```

Comment:
isn't including error object to the log message already printing its stacktrace?

Reply:
done